Error with Registration dates when closing tasks

If you attempt to close a task in Direct and the registration scheduled and end dates are NOT in the future, the app errors complaining that the start date can't be greater than the end date.

The error gets raised on this call:  https://github.com/appirio-tech/direct-app/blob/master/services/contest_service_facade/src/java/main/com/topcoder/service/facade/contest/ejb/ContestServiceFacadeBean.java#L9158

And then triggers the validator in the Phase class:

Caused by: java.lang.IllegalArgumentException: scheduledEndDate can not before scheduledStartDate
	at com.topcoder.project.phases.ProjectPhaseHelper.checkDateNotBefore(ProjectPhaseHelper.java:79
I believe this happens if you create the task, but then don't activate/close it right away.  However, I think we should generically check this case when closing tasks.  We probably need to update the logic for setting the dates.  Maybe if the reg dates are already set, we should not attempt to update them.

--- a/services/contest_service_facade/src/java/main/com/topcoder/service/facade/contest/ejb/ContestServiceFacadeBean.java
+++ b/services/contest_service_facade/src/java/main/com/topcoder/service/facade/contest/ejb/ContestServiceFacadeBean.java
@@ -9106,12 +9106,16 @@ public class ContestServiceFacadeBean implements ContestServiceFacadeLocal, Cont
 
             com.topcoder.project.phases.Project projectPhases = projectServices.getPhases(contest.getId());
             com.topcoder.project.phases.Phase[] phases = projectPhases.getAllPhases();
-            boolean needToUpdate = false;
+            boolean phaseNeedToUpdate = false;
+            boolean phaseHasClosed = false;
             //make sure we have submission phase is opened if not update it
             for (com.topcoder.project.phases.Phase phase : phases) {
                 if (PROJECT_SUBMISSION_PHASE_NAME.equals(phase.getPhaseType().getName())) {
                     if (phase.getPhaseStatus().getId() != PhaseStatus.OPEN.getId()) {
-                        needToUpdate = true;
+                        phaseNeedToUpdate = true;
+                        if (phase.getPhaseStatus().getId() == PhaseStatus.CLOSED.getId()){
+                            phaseHasClosed = true;
+                        }
                         phase.setPhaseStatus(PhaseStatus.OPEN);
                     }
                 } else if (PROJECT_ITERATIVE_REVIEW_PHASE_NAME.equals(phase.getPhaseType().getName()) ||
@@ -9130,7 +9134,7 @@ public class ContestServiceFacadeBean implements ContestServiceFacadeLocal, Cont
                     }
                 }
             }
-            if (needToUpdate) {
+            if (phaseNeedToUpdate) {
                 projectPhases.setPhases(new HashSet<com.topcoder.project.phases.Phase>(Arrays.asList(phases)));
                 projectServices.updatePhases(projectPhases, String.valueOf(tcSubject.getUserId()));
             }
@@ -9145,42 +9149,28 @@ public class ContestServiceFacadeBean implements ContestServiceFacadeLocal, Cont
             com.topcoder.project.phases.Phase submissionPhase = null;
             com.topcoder.project.phases.Phase reviewPhase = null;
             for (com.topcoder.project.phases.Phase phase : phases) {
-                if (PROJECT_REGISTRATION_PHASE_NAME.equals(phase.getPhaseType().getName()) ||
-                        PROJECT_SUBMISSION_PHASE_NAME.equals(phase.getPhaseType().getName()) ||
-                        PROJECT_REVIEW_PHASE_NAME.equals(phase.getPhaseType().getName()) ||
+                if (PROJECT_SUBMISSION_PHASE_NAME.equals(phase.getPhaseType().getName())) {
+                    if (phaseNeedToUpdate && !phaseHasClosed) {
+                        //submission is scheduled
+                        phase.setActualStartDate(new Date());
+                        phase.setActualEndDate(new Date());
+                    } else if (!phaseNeedToUpdate){
+                        //phase already open
+                        phase.setActualEndDate(new Date());
+                    }
+                    phase.setPhaseStatus(PhaseStatus.CLOSED);
+                    submissionPhase = phase;
+                } else if (PROJECT_REVIEW_PHASE_NAME.equals(phase.getPhaseType().getName()) ||
                         PROJECT_ITERATIVE_REVIEW_PHASE_NAME.equals(phase.getPhaseType().getName())) {
-                    Date currentDate = new Date();
-                    //length 1 hour
-                    long length = 60 * MINUTE_IN_MILIS;
-                    //reg start 3h before
-                    Date regStartDate = new Date(currentDate.getTime() - 180 * MINUTE_IN_MILIS);
-                    //submision start 2h before
-                    Date submissionStartDate = new Date(currentDate.getTime() - 120 * MINUTE_IN_MILIS);
-                    //submission end / review start 1h before
-                    Date submissionEndDate = new Date(currentDate.getTime() - 60 * MINUTE_IN_MILIS);
-
-                    if (PROJECT_SUBMISSION_PHASE_NAME.equals(phase.getPhaseType().getName())) {
-                        phase.setScheduledStartDate(submissionStartDate);
-                        phase.setActualStartDate(submissionStartDate);
-                        phase.setScheduledEndDate(submissionEndDate);
-                        phase.setActualEndDate(submissionEndDate);
-                        phase.setLength(length);
-                        phase.setPhaseStatus(PhaseStatus.CLOSED);
-                        submissionPhase = phase;
-                    } else if (PROJECT_REGISTRATION_PHASE_NAME.equals(phase.getPhaseType().getName())) {
-                        phase.setScheduledStartDate(regStartDate);
-                        phase.setFixedStartDate(regStartDate);
-                        phase.setActualStartDate(regStartDate);
-                        phase.setScheduledEndDate(submissionEndDate);
-                        phase.setActualEndDate(submissionEndDate);
-                        phase.setLength(2 * length);
-                        phase.setPhaseStatus(PhaseStatus.CLOSED);
-                    } else {
-                        phase.setScheduledStartDate(submissionEndDate);
-                        phase.setActualStartDate(submissionEndDate);
-                        phase.setScheduledEndDate(currentDate);
-                        phase.setLength(length);
+                    if (phase.getPhaseStatus().getId() == PhaseStatus.SCHEDULED.getId()) {
+                        phase.setActualStartDate(new Date());
+                        phase.setScheduledEndDate(null);
+                        phase.setScheduledStartDate(phase.calcStartDate());
+                        phase.setScheduledEndDate(phase.calcEndDate());
                         phase.setPhaseStatus(PhaseStatus.OPEN);
+                    }
+                    if (phase.getPhaseStatus().getId() != PhaseStatus.CLOSED.getId()) {
+                        //skiping closed iterative review
                         reviewPhase = phase;
                     }
                 }

--- a/services/project_services/src/java/main/com/topcoder/project/service/impl/ProjectServicesImpl.java
+++ b/services/project_services/src/java/main/com/topcoder/project/service/impl/ProjectServicesImpl.java
@@ -2083,17 +2083,10 @@ public class ProjectServicesImpl implements ProjectServices {
                     if (p.getPhaseType().getId() == PhaseType.REGISTRATION_PHASE.getId() && projectHeader.getProjectStatus().getId() == ProjectStatus.DRAFT.getId()) {
                         if (projectHeader.getProperty(ProjectPropertyType.PRIVATE_PROJECT) != null && "1".equals(projectHeader.getProperty(ProjectPropertyType.PRIVATE_PROJECT))) {
                             if (p.getPhaseStatus().getId() != PhaseStatus.CLOSED.getId()) {
-                                Date now = new Date();
-                                p.setFixedStartDate(now);
-                                p.setActualStartDate(now);
-                                p.setActualEndDate(new Date(System.currentTimeMillis()+5*60*1000));
                                 p.setPhaseStatus(PhaseStatus.CLOSED);
                             }
                         } else {
                             if (p.getPhaseStatus().getId() == PhaseStatus.CLOSED.getId()) {
-                                p.setActualStartDate(null);
-                                p.setActualEndDate(null);
-                                p.setFixedStartDate(null);
                                 p.setPhaseStatus(PhaseStatus.SCHEDULED);
                             }
                         }
@@ -5171,10 +5164,6 @@ public class ProjectServicesImpl implements ProjectServices {
             if (p.getPhaseType().getId() == PhaseType.REGISTRATION_PHASE.getId()) {
                 if (projectHeader.getProperty(ProjectPropertyType.PRIVATE_PROJECT) != null && projectHeader.getProperty(ProjectPropertyType.PRIVATE_PROJECT).equals("1") && projectHeader.getProjectStatus().getId() == ProjectStatus.DRAFT.getId()) {
                     p.setPhaseStatus(PhaseStatus.CLOSED);
-                    Date now = new Date();
-                    p.setFixedStartDate(now);
-                    p.setActualStartDate(now);
-                    p.setActualEndDate(new Date(System.currentTimeMillis()+5*60*1000));
                 }
             }
         }

